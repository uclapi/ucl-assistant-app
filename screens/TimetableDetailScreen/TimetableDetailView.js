import React from "react";
import { StyleSheet, View } from "react-native";
import PropTypes from "prop-types";
import { MapView } from "expo";
import moment from "moment";
import {
  ErrorText,
  TitleText,
  SubtitleText,
  BodyText,
  Link,
} from "../../components/Typography";
import Button from "../../components/Button";
import { Page } from "../../components/Containers";
import MapStyle from "../../styles/Map";
import MapsManager from "../../lib/MapsManager";
import MailManager from "../../lib/MailManager";

const styles = StyleSheet.create({
  contactPerson: {
    marginBottom: 20,
    marginTop: 20,
  },
  emailButton: {
    marginTop: 10,
  },
});

class TimetableDetailView extends React.Component {
  navigateToLocation = ({ lat, lng, address }) => () => {
    if (lat && lng) {
      MapsManager.navigateToCoords({ lat, lng });
    } else {
      MapsManager.navigateToAddress(address.join());
    }
  };

  sendEmail = email => () => {
    MailManager.composeAsync({
      recipients: [email],
    });
  };

  openRoomSearch = roomName => () => {
    const { navigation } = this.props;
    navigation.navigate("Rooms", { query: roomName });
  };

  render() {
    let contactTypeStr = "";
    const {
      date,
      location,
      initialRegion,
      contact: contactPerson,
      module,
      session_type_str: sessionTypeStr,
      session_group: sessionGroup,
      start_time: startTime,
      end_time: endTime,
    } = this.props;
    const sessionType = sessionTypeStr.toLowerCase();
    switch (true) {
      case sessionType === "lecture":
      case sessionType === "seminar": {
        contactTypeStr = "Lecturer";
        break;
      }
      case sessionType === "practical":
      case sessionType === "problem based learning": {
        contactTypeStr = "Instructor";
        break;
      }
      default: {
        contactTypeStr = "Contact";
      }
    }

    const { lat, lng } = location.coordinates;
    const latitude = parseFloat(lat, 10) || initialRegion.latitude;
    const longitude = parseFloat(lng, 10) || initialRegion.longitude;
    const { address, type: locationType, name: locationName } = location;

    const { department_name: departmentName, email } = module.lecturer;

    return (
      <Page>
        <TitleText>{module.name}</TitleText>
        <BodyText>{moment(date).format("dddd, Do MMMM YYYY")}</BodyText>
        <BodyText>
          {startTime} - {endTime}
        </BodyText>
        {locationType === "CB" ? (
          <Link onPress={this.openRoomSearch(locationName)}>
            {locationName}
          </Link>
        ) : (
          <BodyText>{locationName}</BodyText>
        )}
        <BodyText>Type: {sessionTypeStr}</BodyText>
        {sessionGroup && sessionGroup.length > 0 ? (
          <BodyText>Group {sessionGroup}</BodyText>
        ) : null}
        {(!lat || !lng) && (
          <ErrorText>
            Error: We couldn{"'"}t fetch coordinates for this venue, so the map
            may be incorrect.
          </ErrorText>
        )}
        <MapView
          style={MapStyle.wideMap}
          initialRegion={initialRegion}
          region={{
            latitude,
            longitude,
            latitudeDelta: initialRegion.latitudeDelta,
            longitudeDelta: initialRegion.longitudeDelta,
          }}
        >
          <MapView.Marker coordinate={{ latitude, longitude }} />
        </MapView>
        <Button
          onPress={() => {
            if (lat && lng) {
              MapsManager.navigateToCoords({ lat, lng });
            } else {
              MapsManager.navigateToAddress(address.join());
            }
          }}
        >
          Directions
        </Button>

        <View style={styles.contactPerson}>
          <SubtitleText>{contactTypeStr}</SubtitleText>
          {contactPerson && contactPerson.length > 0 ? (
            <BodyText>
              {contactPerson}{" "}
              {departmentName && departmentName.length > 0
                ? `(${departmentName})`
                : null}
            </BodyText>
          ) : null}
          {email && email.length > 0 ? (
            <Button onPress={this.sendEmail(email)} style={styles.emailButton}>
              Send Email
            </Button>
          ) : null}
        </View>

        {__DEV__ && <SubtitleText>Debug Information</SubtitleText>}
        {__DEV__ && <BodyText>{JSON.stringify(this.props, "\n", 2)}</BodyText>}
      </Page>
    );
  }
}

TimetableDetailView.propTypes = {
  initialRegion: PropTypes.shape({
    latitude: PropTypes.number,
    longitude: PropTypes.number,
    latitudeDelta: PropTypes.number,
    longitudeDelta: PropTypes.number,
  }),
  date: PropTypes.string,
  start_time: PropTypes.string,
  end_time: PropTypes.string,
  contact: PropTypes.string,
  location: PropTypes.shape({
    name: PropTypes.string,
    address: PropTypes.arrayOf(PropTypes.string),
    coordinates: PropTypes.shape({
      lat: PropTypes.string,
      lng: PropTypes.string,
    }),
    type: PropTypes.string,
  }),
  module: PropTypes.shape({
    name: PropTypes.string,
    lecturer: PropTypes.shape({
      department_id: PropTypes.string,
      department_name: PropTypes.string,
      email: PropTypes.string,
      name: PropTypes.string,
    }),
  }),
  session_type_str: PropTypes.string,
  session_group: PropTypes.string,
  navigation: PropTypes.shape().isRequired,
};

TimetableDetailView.defaultProps = {
  initialRegion: {
    latitude: 51.5246586,
    longitude: -0.1339784,
    latitudeDelta: 0.0012,
    longitudeDelta: 0.0071,
  },
  date: "2019-01-01",
  start_time: "",
  end_time: "",
  contact: "",
  location: {
    name: "",
    address: [],
    coordinates: {
      lat: "51.5246586",
      lng: "-0.1339784",
    },
    type: "DB",
  },
  module: {
    name: "",
    lecturer: {
      department_id: "",
      department_name: "",
      email: "",
      name: "",
    },
  },
  session_type_str: "",
  session_group: "",
};

export default TimetableDetailView;
