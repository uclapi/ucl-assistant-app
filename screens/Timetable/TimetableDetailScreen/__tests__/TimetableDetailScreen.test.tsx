/**
 * @jest-environment jsdom
 */
import "react-native"

import React from 'react'
import ShallowRenderer from 'react-test-renderer/shallow'

import { TimetableDetailScreen } from '../TimetableDetailScreen'

describe(`TimetableDetailScreen`, () => {
  const renderer = new ShallowRenderer()

  const mockProps = {
    navigation: {
      navigate: jest.fn(),
    },
    route: {
      params: {
        code: `ABC123`,
        date: `2020-01-01`,
        time: `00:00`,
      },
    },
    timetable: {
      "2020-01-01": {
        timetable: [
          {
            contact: `Anne Oldman`,
            end_time: `00:01`,
            location: {
              address: [
                `221C Banana Bread Street`,
              ],
              name: `Waffle House`,
            },
            module: {
              department_name: `Criminology`,
              lecturer: {
                department_id: `CRIM`,
                name: `Jack Cloth`,
              },
              module_id: `ABC123`,
            },
            start_time: `00:00`,
          },
        ],
      },
    },
  }

  it(`renders the TimetableDetailScreen`, () => {
    renderer.render(<TimetableDetailScreen {...mockProps} />)
    const result = renderer.getRenderOutput()
    expect(result).toMatchSnapshot()
  })
})
