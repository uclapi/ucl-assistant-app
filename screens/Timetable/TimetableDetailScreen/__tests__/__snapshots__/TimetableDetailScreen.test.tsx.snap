// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimetableDetailScreen renders the TimetableDetailScreen 1`] = `
<TimetableDetailView
  contact=""
  date="2020-01-01"
  end_time=""
  initialRegion={
    Object {
      "latitude": 51.5246586,
      "latitudeDelta": 0.0012,
      "longitude": -0.1339784,
      "longitudeDelta": 0.0071,
    }
  }
  location={
    Object {
      "address": Array [],
      "coordinates": Object {
        "lat": "51.5246586",
        "lng": "-0.1339784",
      },
      "name": "",
      "type": "DB",
    }
  }
  module={
    Object {
      "department_name": "",
      "email": "",
      "lecturer": Object {
        "department_id": "",
        "department_name": "",
        "email": "",
        "name": "",
      },
      "name": "",
    }
  }
  navigation={
    Object {
      "navigate": [MockFunction],
    }
  }
  session_group=""
  session_type_str=""
  start_time=""
/>
`;
